src/IO/RemoteReadBufferCache.cpp

```diff
@@ -0,0 +1,575 @@
+            before_unflush_bytes = 0;
+        }
+    }
+    std::unique_lock lock(mutex);
+    current_offset += total_bytes;
+    file_meta_data.status = RemoteFileMetadata::DOWNLOADED;
+    flush(true);
+    out_file->close();
+    out_file.reset();
+    remote_read_buffer.reset();
+    lock.unlock();
+    more_data_signal.notify_all();
+    RemoteReadBufferCache::instance().updateTotalSize(file_meta_data.file_size);
+    LOG_TRACE(log, "Finish download into local path: {}, file meta data:{} ", local_path.string(), file_meta_data.toString());
+}
+
+void RemoteCacheController::flush(bool need_flush_meta_data_)
+{
+    if (out_file)
+    {
+        out_file->flush();
+    }
+
+    if (!need_flush_meta_data_)
+        return;
+
+    file_meta_data.save(local_path / "meta.txt");
+}
+
+RemoteCacheController::~RemoteCacheController() = default;
+
+void RemoteCacheController::close()
+{
+    // delete directory
+    LOG_TRACE(log, "Removing all local cache. local path: {}, file meta data:{}", local_path.string(), file_meta_data.toString());
+    std::filesystem::remove_all(local_path);
+}
+
+std::pair<FILE *, std::filesystem::path> RemoteCacheController::allocFile()
+{
+    std::filesystem::path result_local_path = local_path / "data.bin";
+
+    FILE * fs = fopen(result_local_path.string().c_str(), "r");
+    if (!fs)
+        throw Exception(ErrorCodes::BAD_GET, "Alloc file failed, error code: {} local path: {}", errno, local_path.string());
+
+    std::lock_guard lock{mutex};
+    opened_file_streams.insert(fs);
+    return {fs, result_local_path};
+}
+
+void RemoteCacheController::deallocFile(FILE * file_stream)
+{
+    {
+        std::lock_guard lock{mutex};
+        auto it = opened_file_streams.find(file_stream);
+        if (it == opened_file_streams.end())
+            throw Exception(
+                ErrorCodes::BAD_ARGUMENTS,
+                "Try to deallocate file with invalid handler remote path: {}, local path: {}",
+                file_meta_data.remote_path,
+                local_path.string());
+        opened_file_streams.erase(it);
+    }
+    fclose(file_stream);
+}
+
+LocalCachedFileReader::LocalCachedFileReader(RemoteCacheController * cache_controller_)
+    : cache_controller(cache_controller_)
+    , file_size(cache_controller_->getFileMetaData().file_size)
+    , offset(0)
+{
+    std::tie(file_stream, local_path) = cache_controller->allocFile();
+}
+
+LocalCachedFileReader::~LocalCachedFileReader()
+{
+    cache_controller->deallocFile(file_stream);
+}
+
+size_t LocalCachedFileReader::read(char * buf, size_t size)
+{
+    auto wret = cache_controller->waitMoreData(offset, offset + size);
+    if (wret != RemoteReadBufferCacheError::OK)
+        return 0;
+    std::lock_guard lock(mutex);
+    auto ret_size = fread(buf, 1, size, file_stream);
+    offset += ret_size;
+    return ret_size;
+}
+
+off_t LocalCachedFileReader::seek(off_t off)
+{
+    cache_controller->waitMoreData(off, 0);
```

Comment:
I made a comment about this somewhere above, but there are so many comments so it's hard to find your answer now.
So I'll repeat.
I think that we do not need to call `waitMoreData` in `seek` method. It looks redundant.

Reply:
I am trying to use ReadBufferFromFileBase instead. Not sure it works or not, since it crashes now

Reply:
Interesting... How does it crash?

Reply:
we make a new implementation on ReadBufferFromFile